### PR TITLE
build: Use `buf` to compile Protobufs

### PR DESCRIPTION
**Pull request by judahrand**
**What this PR does / why we need it**:

I stumbled across `buf` (https://buf.build) which looks like a really useful tool for making compiling Protobufs much easier. Thought I'd try integrating it into the Feast build process.

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -210,13 +205,9 @@ class BuildPythonProtosCommand(Command):
     ]
 
     def initialize_options(self):
-        self.python_protoc = [
-            sys.executable,
-            "-m",
-            "grpc_tools.protoc",
-        ]  # find_executable("protoc")
+        self.buf = [find_executable("buf")]
```
Comment (achals) [inline setup.py:208]: How is `buf` dependency pulled in? Does it have to be `brew install`ed or somthing?

**Comment by achals** [comment]
Thanks for the PR @judahrand ! Funny timing, I also stumbled on buf and thought it looked cool, so thanks for trying to integrate it here!

File: setup.py (hunk 1 above)
Comment (judahrand) [reply setup.py:208]: I has a `brew install` or can `go install` it

https://docs.buf.build/installation

**Comment by judahrand** [comment]
> Thanks for the PR @judahrand ! Funny timing, I also stumbled on buf and thought it looked cool, so thanks for trying to integrate it here!

The remote Protobuf compilation is especially handy! Looks like it can remove a ton of boiler plate dependency management / installation.

**Comment by judahrand** [comment]
There's still some tidy up but this seems to be loosely working

**Comment by pyalex** [comment]
Sorry for the dumb question, but what problem is being solved here?

**Comment by judahrand** [comment]
> Sorry for the dumb question, but what problem is being solved here?

No particular specific problem in this project. But it does offload all the complexity of compiling the Protobuf definitions into code on multiple languages to a single tool / config location and does away with some of the logic around protoc plugins. More of this logic can be removed than I have so far I think. 

The buf tool also has some additional nice functionality such as linting and formatting of the Protobuf files.

File: protos/buf.python.gen.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+version: v1
```
[the remark this answers is not in the document]
Comment (judahrand) [reply protos/buf.python.gen.yaml:1]: Yeah, so far I've only looked at what was handled in `setup.py` but I see no reason why any Java classes couldn't be produced in the same way.

[the remark this answers is not in the document]
Comment (judahrand) [reply protos/buf.python.gen.yaml:1]: Looks like the Java Protobuf compilation is handled by Maven at the mo'.

**Pull request closed without merging**